### ISSUE TITLE
C++ wrapper object binding never getting released

### ISSUE THREAD

**Issue by codewarrior0** [PySide/shiboken2#3]
Hi! Sorry if this is a little sparse, but I'm having trouble holding it all in my head at once.

I am getting this completely bizarre error:

```
 item = QtGui.QStandardItem(clean_msg(record.getMessage()))
>item.setForeground(QtGui.QBrush(fg))

AttributeError: 'PySide.QtOpenGL.QGLContext' object has no attribute 'setData'
```

I didn't call `setData`. Why is it calling `setData` for me? And why is there a `QGLContext` involved? I spent at least a dozen hours (and constructed more than one hundred million QStandardItems) to arrive at this answer: The answer is because of the "C++ Wrapper" classes, which are the subclasses of C++ classes that allow the virtual methods of that class to be reimplemented by the python wrapper.

The Python wrapper wraps the C++ wrapper, which subclasses the underlying C++ class. How does the C++ wrapper know which Python wrapper to call? It asks the BindingManager for a mapping from the C++ wrapper to the Python wrapper.

When an object is instantiated from Python, it also instantiates a C++ wrapper, and the Python object wraps the C++ wrapper. It also goes to the BindingManager and registers the two wrappers as a mated pair. 

This ceremony is done [in cppgenerator.cpp:1397](https://github.com/PySide/shiboken2/blob/qt5/generator/shiboken/cppgenerator.cpp#L1397). But something seems funny about those two lines.

If the wrapped class is a C++ wrapper, then it sets `hasCppWrapper` on the SbkObject to True. However, even if the wrapped class is *not* a C++ wrapper, it goes ahead and registers it with the BindingManager anyway. This is a no-op with respect to virtual functions; the BindingManager will never be asked for the Python wrapper corresponding to a C++ object that is not a C++ wrapper.

Here is where the problem lies: [In basewrapper.cpp:832](https://github.com/PySide/shiboken2/blob/qt5/libshiboken/basewrapper.cpp#L832), the binding will be released from the BindingManager *if and only if* the SbkObject's `hasCppWrapper` is true. 

So, the Python object is deallocated, the C++ object it wrapped was deallocated, but the BindingManager still has a dangling mapping between the two vacated memory addresses. I hope you can see where this is going.

A new QStandardItem is created at the address vacated by the old C++ object. Since this QStandardItem was constructed in Python code, it creates a C++ wrapper for the QStandardItem instead. A new SbkObject for a QGLContext is created at the address vacated by the old SbkObject. I call setForeground on the QStandardItem, and the implementation of `setForeground` calls the virtual function `setData`.

The BindingManager still thinks the two addresses are mated, so when Qt calls that virtual function on the QStandardItem (remember, it is a C++ wrapper), the wrapper tries to forward the virtual function call to its Python object, but finds a QGLContext instead - thus raising the AttributeError.

The obvious fix, I hope, is to [add some curly braces here](https://github.com/PySide/shiboken2/blob/qt5/generator/shiboken/cppgenerator.cpp#L1397) to make sure that the binding is registered *if and only if* the SbkObject is wrapping a C++ wrapper.

**Comment by codewarrior0** [comment]
Oops, I almost forgot.

This bug is present in PySide 1.2.2, and was discovered on Python 2.7.10 for 64-bit Windows.

**Comment by ctismer** [comment]
Ok, then you probably don't want to put the issue here and now, because
this is just for the new Qt5 version. I think you are better off with the original
shiboken issue tracker, unless you want to help me with PySide2 :-)

**Comment by codewarrior0** [comment]
Seeing as the PySide 1 repos haven't moved in 4 years I figured here was as good a place as any. I'm not holding my breath for a 1.2.3 release.

The source lines I linked seem to say the same thing in both 1 and 2, so the bug is probably still around. They are not the only places `BindingManager::registerWrapper` and `BindingManager::releaseWrapper` are called, either, so I don't know if what I suggested is a complete fix.

**Comment by ctismer** [comment]
Again: The work on shiboken 1.2.3 does not happen here. It is only there for reference.
Development happens in the Qt5 branch, and this starts with 2.0.0.dev0.
When the master branch of shiboken gets a fix, then I will merge it.
But here you are on a dead end, 1.2.2-wise ;-)

**Comment by codewarrior0** [comment]
Well, the code is the same on 2.0.0dev0, so I would expect you to have the same issue there too. Other than that, I'm not sure what you're trying to tell me. Some comments on the issue itself would be appreciated ;-)

I've had no luck trying to create a minimal test case before, but that's partly because it didn't happen reliably in the app I'm developing, and partly because I didn't understand what was going on underneath. Now that I do understand, I think a test case would have to do something like:

Repeatedly create and destroy lots of objects by directly constructing them from Python, and also repeatedly create lots of objects indirectly by directly constructing some other object from Python that constructs a child object from C++ and then getting a Python wrapper for that child object, then destroying the wrapper. Eventually, the child objects will fill up the BindingManager with lots of mappings for dead objects, and you'll end up with a Python wrapper, wrapping a C++ wrapper, that is bound to a Python wrapper other than the one that actually wraps it.

**Comment by codewarrior0** [comment]
Never mind.

**Comment by ctismer** [comment]
Ok, interesting, and something that I observed sometimes as well.

I had a look at your patch, and what you say is completely right.
I will merge that into 2.0, of course.

You should duplicate your issue for the older shiboken repo, because
most developers are still subscribed there.

Sorry about my hasty reaction without reading. I am swamped with work and feared
even more work.

Thanks for your help !

**Comment by codewarrior0** [comment]
Thanks for reading. I understand why it would be intimidating, there's a lot to keep in mind at once, and it would be nearly impossible if you're preoccupied with other work.

I spent some more time reading a lot of shiboken's code to try to get a clearer picture. Now, I don't think my proposed fix is right at all. I thought `BindingManager` was only used for CppWrappers, but that's not the case - it's used to map *any* C++ object to its Python wrapper.

This is important because, when you do something like `myGlWidget.context()`, the generated function calls `newObject()`, which asks the `BindingManager` to find the existing Python wrapper for that object. I confirmed that you do get the same Python wrapper each time you call `myGlWidget.context()`, so this part is working as intended.

The best explanation I can come up with for the bizarre error at the beginning is that old mappings registered with the `BindingManager` aren't being released for some reason. The asymmetry I noticed in the two source lines I linked is explained by how `invalidate` is responsible for releasing the binding in the case that the C++ object is *not* a CppWrapper; while `callCppDestructors` has that responsibility if it *is* a CppWrapper.

I still have had no success creating a minimal test case, but I've modified my application in such a way that I can get the bizarre error 90% of the time.

I'm now able to actually *build* PySide 1.2.2, so I'm in a position to actually test my imagined fixes instead of bombarding a poor, overworked developer with mad speculations and expecting him to read all of them. I'll get back to you when I can come up with something that works.

**Comment by ctismer** [comment]
So I should _not_ apply the patch?
Well, shiboken is quite a buggy thing, partially, and I'm very much considering to
re-write it in a better language and improve it.

Right now I am fixing PySide problems in the moving to Qt5, which has made some
substantial changes, which I need to adapt to:

One thing is qmetaobject.cpp which has changed the supported revisions.
PySide still uses revision 3, but Qt5 enforces revision 7, and I am trying hard
to understand what I have to change in libpyside/dynamicqmetaobject.cpp

I would very much appreciate if you can help a bit.

**Comment by codewarrior0** [comment]
No, I think the patch I proposed would actually break quite a few things. 

If I can come up with a patch that actually fixes the problem, and doesn't break anything along the way, I'll send you a PR on the other repo.

**Comment by ctismer** [comment]
@codewarrior0 can you please get back to this? I don't know how to cope with that.
